docker-java push fails against the podman service

<!--
---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------
Use the commands below to provide key information from your environment:
You do NOT have to include this information if this is a FEATURE REQUEST

**NOTE** A large number of issues reported against Podman are often found to already be fixed
in more current versions of the project.  Before reporting an issue, please verify the
version you are running with `podman version` and compare it to the latest release
documented on the top of Podman's [README.md](../README.md).  If they differ, please
update your version of Podman to the latest possible and retry your command before creating
an issue.

Also, there is a running list of known issues in the [Podman Troubleshooting Guide](https://github.com/containers/podman/blob/master/troubleshooting.md),
please reference that page before opening a new issue.

If you are filing a bug against `podman build`, please instead file a bug
against Buildah (https://github.com/containers/buildah/issues). Podman build
executes Buildah to perform container builds, and as such the Buildah
maintainers are best equipped to handle these bugs.
-->

**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

<!--
Briefly describe the problem you are having in a few paragraphs.
-->

**Steps to reproduce the issue:**

1. start podman as a system service: `podman system service tcp:0.0.0.0:2375 --timo=0`

2. wrote a gradle build script that uses https://github.com/bmuschko/gradle-docker-plugin to build docker image and to push it

3. the build fails during the push

**Describe the results you received:**
The push fails with:
```
Caused by: com.github.dockerjava.api.exception.DockerClientException: Could not push image
	at com.github.dockerjava.core.command.PushImageResultCallback.throwFirstError(PushImageResultCallback.java:40)
	at com.github.dockerjava.api.async.ResultCallbackTemplate.awaitCompletion(ResultCallbackTemplate.java:93)
	at com.bmuschko.gradle.docker.tasks.image.DockerPushImage.runRemoteCommand(DockerPushImage.groovy:63)
	at com.bmuschko.gradle.docker.tasks.AbstractDockerRemoteApiTask.start(AbstractDockerRemoteApiTask.groovy:74)
```

**Describe the results you expected:**
The push to work as podman's API implementation should be compatible to that of docker.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
Version:      2.2.1
API Version:  2.1.0
Go Version:   go1.15.5
Built:        Tue Dec  8 14:38:17 2020
OS/Arch:      linux/s390x
```

**Output of `podman info --debug`:**

```
host:
  arch: s390x
  buildahVersion: 1.18.0
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: conmon-2.0.21-3.fc33.s390x
    path: /usr/bin/conmon
    version: 'conmon version 2.0.21, commit: d47b0c4859cb71c991115d47c69e732fe8e5d30c'
  cpus: 8
  distribution:
    distribution: fedora
    version: "33"
  eventLogger: file
  hostname: podman-host-0
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 4.18.0-147.el8.s390x
  linkmode: dynamic
  memFree: 401321984
  memTotal: 16868089856
  ociRuntime:
    name: crun
    package: crun-0.16-1.fc33.s390x
    path: /usr/bin/crun
    version: |-
      crun version 0.16
      commit: eb0145e5ad4d8207e84a327248af76663d4e50dd
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  remoteSocket:
    path: /run/podman/podman.sock
  rootless: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 0
  swapTotal: 0
  uptime: 338h 44m 58.82s (Approximately 14.08 days)
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - registry.centos.org
  - docker.io
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.imagestore: /var/lib/shared
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: fuse-overlayfs-1.3.0-1.fc33.s390x
      Version: |-
        fusermount3 version: 3.9.3
        fuse-overlayfs: version 1.3
        FUSE library version 3.9.3
        using FUSE kernel interface version 7.31
    overlay.mountopt: nodev,fsync=0
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: overlayfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 3
  runRoot: /var/run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 2.1.0
  Built: 1607438297
  BuiltTime: Tue Dec  8 14:38:17 2020
  GitCommit: ""
  GoVersion: go1.15.5
  OsArch: linux/s390x
  Version: 2.2.1
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-2.2.1-1.fc33.s390x
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**
A pod in an OpenShift setup.

--- a/libpod/image/image.go
+++ b/libpod/image/image.go
@@ -816,7 +816,7 @@ func (i *Image) UntagImage(tag string) error {
 
 // PushImageToHeuristicDestination pushes the given image to "destination", which is heuristically parsed.
 // Use PushImageToReference if the destination is known precisely.
-func (i *Image) PushImageToHeuristicDestination(ctx context.Context, destination, manifestMIMEType, authFile, digestFile, signaturePolicyPath string, writer io.Writer, forceCompress bool, signingOptions SigningOptions, dockerRegistryOptions *DockerRegistryOptions, additionalDockerArchiveTags []reference.NamedTagged) error {
+func (i *Image) PushImageToHeuristicDestination(ctx context.Context, destination, manifestMIMEType, authFile, digestFile, signaturePolicyPath string, writer io.Writer, forceCompress bool, signingOptions SigningOptions, dockerRegistryOptions *DockerRegistryOptions, additionalDockerArchiveTags []reference.NamedTagged, progress chan types.ProgressProperties) error {
 	if destination == "" {
 		return errors.Wrapf(syscall.EINVAL, "destination image name must be specified")
 	}
@@ -834,11 +834,11 @@ func (i *Image) PushImageToHeuristicDestination(ctx context.Context, destination
 			return err
 		}
 	}
-	return i.PushImageToReference(ctx, dest, manifestMIMEType, authFile, digestFile, signaturePolicyPath, writer, forceCompress, signingOptions, dockerRegistryOptions, additionalDockerArchiveTags)
+	return i.PushImageToReference(ctx, dest, manifestMIMEType, authFile, digestFile, signaturePolicyPath, writer, forceCompress, signingOptions, dockerRegistryOptions, additionalDockerArchiveTags, progress)
 }
 
 // PushImageToReference pushes the given image to a location described by the given path
-func (i *Image) PushImageToReference(ctx context.Context, dest types.ImageReference, manifestMIMEType, authFile, digestFile, signaturePolicyPath string, writer io.Writer, forceCompress bool, signingOptions SigningOptions, dockerRegistryOptions *DockerRegistryOptions, additionalDockerArchiveTags []reference.NamedTagged) error {
+func (i *Image) PushImageToReference(ctx context.Context, dest types.ImageReference, manifestMIMEType, authFile, digestFile, signaturePolicyPath string, writer io.Writer, forceCompress bool, signingOptions SigningOptions, dockerRegistryOptions *DockerRegistryOptions, additionalDockerArchiveTags []reference.NamedTagged, progress chan types.ProgressProperties) error {
 	sc := GetSystemContext(signaturePolicyPath, authFile, forceCompress)
 	sc.BlobInfoCacheDir = filepath.Join(i.imageruntime.store.GraphRoot(), "cache")
 
@@ -859,6 +859,10 @@ func (i *Image) PushImageToReference(ctx context.Context, dest types.ImageRefere
 	}
 	copyOptions := getCopyOptions(sc, writer, nil, dockerRegistryOptions, signingOptions, manifestMIMEType, additionalDockerArchiveTags)
 	copyOptions.DestinationCtx.SystemRegistriesConfPath = registries.SystemRegistriesConfPath() // FIXME: Set this more globally.  Probably no reason not to have it in every types.SystemContext, and to compute the value just once in one place.
+	if progress != nil {
+		copyOptions.Progress = progress
+		copyOptions.ProgressInterval = time.Second
+	}
 	// Copy the image to the remote destination
 	manifestBytes, err := cp.Image(ctx, policyContext, dest, src, copyOptions)
 	if err != nil {
@@ -1648,7 +1652,7 @@ func (i *Image) Save(ctx context.Context, source, format, output string, moreTag
 			return err
 		}
 	}
-	if err := i.PushImageToReference(ctx, destRef, manifestType, "", "", "", writer, compress, SigningOptions{RemoveSignatures: removeSignatures}, &DockerRegistryOptions{}, additionaltags); err != nil {
+	if err := i.PushImageToReference(ctx, destRef, manifestType, "", "", "", writer, compress, SigningOptions{RemoveSignatures: removeSignatures}, &DockerRegistryOptions{}, additionaltags, nil); err != nil {
 		return errors.Wrapf(err, "unable to save %q", source)
 	}
 	i.newImageEvent(events.Save)

--- a/pkg/api/handlers/compat/images_push.go
+++ b/pkg/api/handlers/compat/images_push.go
@@ -1,6 +1,8 @@
 package compat
 
 import (
+	"context"
+	"encoding/json"
 	"fmt"
 	"io/ioutil"
 	"net/http"
@@ -10,11 +12,14 @@ import (
 	"github.com/containers/podman/v3/libpod"
 	"github.com/containers/podman/v3/pkg/api/handlers/utils"
 	"github.com/containers/podman/v3/pkg/auth"
+	"github.com/containers/podman/v3/pkg/channel"
 	"github.com/containers/podman/v3/pkg/domain/entities"
 	"github.com/containers/podman/v3/pkg/domain/infra/abi"
 	"github.com/containers/storage"
+	"github.com/docker/docker/pkg/jsonmessage"
 	"github.com/gorilla/schema"
 	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
 )
 
 // PushImage is the handler for the compat http endpoint for pushing images.
@@ -82,6 +87,8 @@ func PushImage(w http.ResponseWriter, r *http.Request) {
 		Password:   password,
 		Username:   username,
 		DigestFile: digestFile.Name(),
+		Quiet:      true,
+		Progress:   make(chan types.ProgressProperties),
 	}
 	if _, found := r.URL.Query()["tlsVerify"]; found {
 		options.SkipTLSVerify = types.NewOptionalBool(!query.TLSVerify)
@@ -94,31 +101,103 @@ func PushImage(w http.ResponseWriter, r *http.Request) {
 		destination = imageName
 	}
 
-	if err := imageEngine.Push(r.Context(), imageName, destination, options); err != nil {
-		if errors.Cause(err) != storage.ErrImageUnknown {
-			utils.ImageNotFound(w, imageName, errors.Wrapf(err, "failed to find image %s", imageName))
-			return
+	errorWriter := channel.NewWriter(make(chan []byte))
+	defer errorWriter.Close()
+
+	statusWriter := channel.NewWriter(make(chan []byte))
+	defer statusWriter.Close()
+
+	runCtx, cancel := context.WithCancel(context.Background())
+	var failed bool
+
+	go func() {
+		defer cancel()
+
+		statusWriter.Write([]byte(fmt.Sprintf("The push refers to repository [%s]", imageName)))
+
+		err := imageEngine.Push(runCtx, imageName, destination, options)
+		if err != nil {
+			if errors.Cause(err) != storage.ErrImageUnknown {
+				errorWriter.Write([]byte("An image does not exist locally with the tag: " + imageName))
+			} else {
+				errorWriter.Write([]byte(err.Error()))
+			}
 		}
+	}()
 
-		utils.Error(w, "Something went wrong.", http.StatusBadRequest, errors.Wrapf(err, "error pushing image %q", imageName))
-		return
+	flush := func() {
+		if flusher, ok := w.(http.Flusher); ok {
+			flusher.Flush()
+		}
 	}
 
-	digestBytes, err := ioutil.ReadAll(digestFile)
-	if err != nil {
-		utils.Error(w, "Something went wrong.", http.StatusInternalServerError, errors.Wrap(err, "failed to read digest tmp file"))
-		return
-	}
+	w.WriteHeader(http.StatusOK)
+	w.Header().Add("Content-Type", "application/json")
+	flush()
 
-	tag := query.Tag
-	if tag == "" {
-		tag = "latest"
-	}
-	respData := struct {
-		Status string `json:"status"`
-	}{
-		Status: fmt.Sprintf("%s: digest: %s size: null", tag, string(digestBytes)),
-	}
+	enc := json.NewEncoder(w)
+	enc.SetEscapeHTML(true)
 
-	utils.WriteJSON(w, http.StatusOK, &respData)
+loop: // break out of for/select infinite loop
+	for {
+		var report jsonmessage.JSONMessage
+
+		select {
+		case e := <-options.Progress:
+			switch e.Event {
+			case types.ProgressEventNewArtifact:
+				report.Status = "Preparing"
+			case types.ProgressEventRead:
+				report.Status = "Pushing"
+				report.Progress = &jsonmessage.JSONProgress{
+					Current: int64(e.Offset),
+					Total:   e.Artifact.Size,
+				}
+			case types.ProgressEventSkipped:
+				report.Status = "Layer already exists"
+			case types.ProgressEventDone:
+				report.Status = "Pushed"
+			}
+			report.ID = e.Artifact.Digest.Encoded()[0:12]
+			if err := enc.Encode(report); err != nil {
+				errorWriter.Write([]byte(err.Error()))
+			}
+			flush()
+		case e := <-statusWriter.Chan():
+			report.Status = string(e)
+			if err := enc.Encode(report); err != nil {
+				errorWriter.Write([]byte(err.Error()))
+			}
+			flush()
+		case e := <-errorWriter.Chan():
+			failed = true
+			report.Error = &jsonmessage.JSONError{
+				Message: string(e),
+			}
+			report.ErrorMessage = string(e)
+			if err := enc.Encode(report); err != nil {
+				logrus.Warnf("Failed to json encode error %q", err.Error())
+			}
+			flush()
+		case <-runCtx.Done():
+			if !failed {
+				digestBytes, err := ioutil.ReadAll(digestFile)
+				if err == nil {
+					tag := query.Tag
+					if tag == "" {
+						tag = "latest"
+					}
+					report.Status = fmt.Sprintf("%s: digest: %s", tag, string(digestBytes))
+					if err := enc.Encode(report); err != nil {
+						logrus.Warnf("Failed to json encode error %q", err.Error())
+					}
+					flush()
+				}
+			}
+			break loop // break out of for/select infinite loop
+		case <-r.Context().Done():
+			// Client has closed connection
+			break loop // break out of for/select infinite loop
+		}
+	}
 }

--- a/pkg/api/handlers/libpod/images.go
+++ b/pkg/api/handlers/libpod/images.go
@@ -451,6 +451,7 @@ func PushImage(w http.ResponseWriter, r *http.Request) {
 		Password: password,
 		Format:   query.Format,
 		All:      query.All,
+		Quiet:    true,
 	}
 	if _, found := r.URL.Query()["tlsVerify"]; found {
 		options.SkipTLSVerify = types.NewOptionalBool(!query.TLSVerify)

--- a/pkg/domain/entities/images.go
+++ b/pkg/domain/entities/images.go
@@ -203,6 +203,8 @@ type ImagePushOptions struct {
 	SignBy string
 	// SkipTLSVerify to skip HTTPS and certificate verification.
 	SkipTLSVerify types.OptionalBool
+	// Progress to get progress notifications
+	Progress chan types.ProgressProperties
 }
 
 // ImageSearchOptions are the arguments for searching images.

--- a/pkg/domain/infra/abi/images.go
+++ b/pkg/domain/infra/abi/images.go
@@ -376,7 +376,8 @@ func (ir *ImageEngine) Push(ctx context.Context, source string, destination stri
 		options.Compress,
 		signOptions,
 		&dockerRegistryOptions,
-		nil)
+		nil,
+		options.Progress)
 	if err != nil && errors.Cause(err) != storage.ErrImageUnknown {
 		// Image might be a manifest list so attempt a manifest push
 		if _, manifestErr := ir.ManifestPush(ctx, source, destination, options); manifestErr == nil {

--- a/test/apiv2/12-imagesMore.at
+++ b/test/apiv2/12-imagesMore.at
@@ -46,6 +46,10 @@ t POST "images/localhost:5000/myrepo/push?tlsVerify=false&tag=mytag" '' 200
 # Untag the image
 t POST "libpod/images/$iid/untag?repo=localhost:5000/myrepo&tag=mytag" '' 201
 
+# Try to push non-existing image
+t POST "images/localhost:5000/idonotexist/push?tlsVerify=false" '' 200
+jq -re 'select(.errorDetail)' <<<"$output" &>/dev/null || echo -e "${red}not ok: error message not found in output${nc}" 1>&2
+
 t GET libpod/images/$IMAGE/json 200 \
   .RepoTags[-1]=$IMAGE
 

